arangod/Aql/Ast.cpp

```diff
@@ -2624,170 +2624,218 @@ size_t Ast::countReferences(AstNode const* node, Variable const* search) {
+        }
 
-      if (variable == nullptr) {
-        THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
-      }
+        // p.vertices
+        //   ^
+        state.seen.pop_back();
+
+        if (state.seen.empty()) {
+          return false;
+        }
+
+        top = state.seen.back();
+
+        if (top->type != NODE_TYPE_INDEXED_ACCESS &&
+            top->type != NODE_TYPE_EXPANSION) {
+          return false;
+        }
 
-      auto [it, emp] =
-          result.try_emplace(variable, arangodb::lazyConstruct([&] {
-                               return std::unordered_set<std::string>(
-                                   {std::string(attributeName, nameLength)});
-                             }));
-      if (emp) {
-        // insert attributeName only
-        (*it).second.emplace(attributeName, nameLength);
+        // p.vertices[0]...
+        //           ^
+        // or
+        //
+        // v[0]...
+        //  ^
+        state.seen.pop_back();
+
+        if (state.seen.empty()) {
+          return false;
+        }
       }
 
-      // fall-through
-    }
+      TRI_ASSERT(!state.seen.empty());
+      top = state.seen.back();
 
-    attributeName = nullptr;
-    nameLength = 0;
-    return true;
-  };
+      if (top->type != NODE_TYPE_ATTRIBUTE_ACCESS) {
+        // something following that we cannot handle
+        return false;
+      }
 
-  traverseReadOnly(node, visitor, ::doNothingVisitor);
+      // now take off all projections from the stack
+      std::vector<std::string> path;
+      while (top->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
+        path.emplace_back(top->getString());
+        state.seen.pop_back();
+        if (state.seen.empty()) {
+          break;
+        }
+        top = state.seen.back();
+      }
 
-  return result;
-}
+      TRI_ASSERT(!path.empty());
+      state.attributes.emplace(std::move(path));
+    }
 
-/// @brief determines the top-level attributes referenced in an expression for
-/// the specified out variable
-bool Ast::getReferencedAttributes(AstNode const* node, Variable const* variable,
-                                  std::unordered_set<std::string>& vars) {
-  // traversal state
-  struct TraversalState {
-    Variable const* variable;
-    char const* attributeName;
-    size_t nameLength;
-    bool isSafeForOptimization;
-    std::unordered_set<std::string>& vars;
+    return true;
   };
 
-  TraversalState state{variable, nullptr, 0, true, vars};
-
-  auto visitor = [&state](AstNode const* node) -> bool {
+  auto visitor = [&](AstNode const* node) -> bool {
     if (node == nullptr || !state.isSafeForOptimization) {
       return false;
     }
 
-    if (node->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
-      state.attributeName = node->getStringValue();
-      state.nameLength = node->getStringLength();
-      return true;
-    }
+    if (node->type == NODE_TYPE_EXPANSION) {
+      // special stunt needed here for the [*] operator...
+      if (node->numMembers() >= 5) {
```

Comment:
That is the API of the EXPANSION Node.
If has some set of members (5) each indicating a different piece of the expansion (e.g. in-line filtering/projection)
This Type based API is hard to reason about I totally agree, refactoring this would be more effort, and I would like to keep this out of this PR.

Reply:
Could there be a comment above the >= 5 that just says that ">=5 means X for the expansion node"?

Reply:
Added